src/ccs/multifolding.rs

```diff
@@ -25,58 +25,106 @@ use core::{cmp::max, marker::PhantomData};
-// instances. Then the rest of CCS, CCCS, LCCCS hold references to it.
-// Is our single source of data.
+/// The NIMFS (Non-Interactive MultiFolding Scheme) structure is the center of operations of the folding scheme.
+/// Once generated, it allows us to fold any upcomming CCCS instances within it without needing to do much.
+// XXX: Pending to add doc examples.
 #[derive(Debug)]
-pub struct Multifolding<G: Group> {
-  ccs: CCSShape<G>,
+pub struct NIMFS<G: Group> {
+  ccs: CCS<G>,
   ccs_mle: Vec<MultilinearPolynomial<G::Scalar>>,
+  ck: CommitmentKey<G>,
+  lcccs: LCCCS<G>,
 }
 
-impl<G: Group> Multifolding<G> {
-  /// Generates a new Multifolding instance based on the given CCS.
-  pub fn new(ccs: CCSShape<G>) -> Self {
-    let ccs_mle = ccs.M.iter().map(|matrix| matrix.to_mle()).collect();
-    Self { ccs, ccs_mle }
+impl<G: Group> NIMFS<G> {
+  /// Generates a new NIMFS instance based on the given CCS instance, it's matrix mle's, an existing LCCCS instance and a commitment key to the CCS.
+  pub fn new(
+    ccs: CCS<G>,
+    ccs_mle: Vec<MultilinearPolynomial<G::Scalar>>,
+    lcccs: LCCCS<G>,
+    ck: CommitmentKey<G>,
+  ) -> Self {
+    Self {
+      ccs,
+      ccs_mle,
+      ck,
+      lcccs,
+    }
+  }
+
+  /// Initializes a NIMFS instance given the CCS of it and a first witness vector that satifies it.
+  // XXX: This should probably return an error as we should check whether is satisfied or not.
+  pub fn init<R: RngCore>(mut rng: &mut R, ccs: CCS<G>, z: Vec<G::Scalar>) -> Self {
+    let ccs_mle: Vec<MultilinearPolynomial<G::Scalar>> =
+      ccs.M.iter().map(|matrix| matrix.to_mle()).collect();
+    let w: Vec<G::Scalar> = z[(1 + ccs.l)..].to_vec();
+    let ck = ccs.commitment_key();
+    let r_w = G::Scalar::random(&mut rng);
+    let w_comm = <G as Group>::CE::commit(&ck, &w);
+
+    let r_x: Vec<G::Scalar> = vec![G::Scalar::random(&mut rng); ccs.s];
+    let v = ccs.compute_v_j(&z, &r_x, &ccs_mle);
+
+    let lcccs: LCCCS<G> = LCCCS::new(&ccs, &ccs_mle, &ck, z, &mut rng);
+
+    Self {
+      ccs,
+      ccs_mle,
+      lcccs,
+      ck,
+    }
+  }
+
+  /// Generates a new [`CCCS`] instance ready to be folded.
+  pub fn new_cccs(&self, z: Vec<G::Scalar>) -> CCCS<G> {
+    CCCS::new(&self.ccs, &self.ccs_mle, z, &self.ck)
+  }
+
+  /// This function checks whether the current IVC after the last fold performed is satisfied and returns an error if it isn't.
+  pub fn is_sat(&self) -> Result<(), NovaError> {
+    self.lcccs.is_sat(&self.ccs, &self.ccs_mle, &self.ck)
   }
-}
 
-impl<G: Group> Multifolding<G> {
-  /// Compute sigma_i and theta_i from step 4
+  /// Compute sigma_i and theta_i from step 4.
   pub fn compute_sigmas_and_thetas(
     &self,
-    z1: &Vec<G::Scalar>,
-    z2: &Vec<G::Scalar>,
+    // Note `z2` represents the input of the incoming CCCS instance.
+    // As the current IVC accumulated input is holded inside of the NIMFS(`self`).
+    z2: &[G::Scalar],
```

Comment:
Here is called `z2` but in other methods such as `init` is using `z` (no `z1`) for what in a older version was `z1`. Since for the running instance it uses `z`, feels weird that for the new instance it `z2` (directly using `2` when there is no `1`).
Maybe we could find some naming for that?

Reply:
I can just say `z` or `incoming_z`?

Reply:
Probably not using `z` is better, to avoid confusion with the running instance's `z` (old `z1`).